x/operator/keeper/abci.go

```diff
@@ -90,10 +107,90 @@ func (k *Keeper) UpdateVotingPower(ctx sdk.Context, avsAddr string) error {
+	// where no snapshot information has been stored. Therefore, it is necessary to store
+	// both the snapshot and the helper information.
+	var snapshotHelper operatortypes.SnapshotHelper
+	if !k.HasSnapshotHelper(cc, avsAddr) {
+		isSnapshotChanged = true
+	} else {
+		snapshotHelper, err = k.GetSnapshotHelper(cc, avsAddr)
+		if err != nil {
+			return err
+		}
+	}
+	votingPowerSnapshot := operatortypes.VotingPowerSnapshot{
+		EpochIdentifier: epochIdentifier,
+		EpochNumber:     epochNumber,
+		BlockTime:       ctx.BlockTime(),
+	}
+
+	// The voting power calculated at the end of the current epoch will be applied
+	// to the next epoch. Therefore, when storing the voting power snapshot, we use
+	// the `start_height` of the next epoch as the key. This ensures that during the
+	// slashing process, there is no need to account for voting power activation delay;
+	// it can be used directly.
+	// Use the current height as the snapshot height when handling snapshots triggered
+	// by slashing. This prevents stakers from escaping slashes through backrunning
+	// undelegation.
+	snapshotHeight := ctx.BlockHeight()
+	if !isForSlash {
+		// clear the slash flag at the end of the epoch
+		snapshotHelper.HasSlash = false
+		// Use the start height of the next epoch as the snapshot key.
+		// The start height of the next epoch should be the current height plus 1,
+		// as voting power is updated at the end of the epoch.
+		snapshotHeight++
+		votingPowerSnapshot.EpochNumber++
+	}
+	if snapshotHelper.HasOptOut || isSnapshotChanged {
+		votingPowerSnapshot.TotalVotingPower = avsVotingPower
+		votingPowerSnapshot.VotingPowerSet = votingPowerSet
+		snapshotHelper.LastChangedHeight = snapshotHeight
+		// clear the hasOptOut flag if it's certain that the snapshot will be updated
+		snapshotHelper.HasOptOut = false
+	}
+	votingPowerSnapshot.LastChangedHeight = snapshotHelper.LastChangedHeight
+
+	err = k.SetSnapshotHelper(cc, avsAddr, &snapshotHelper)
+	if err != nil {
+		return err
+	}
+	snapshotKey := assetstype.GetJoinedStoreKey(avsAddr, strconv.FormatInt(snapshotHeight, 10))
+	err = k.SetVotingPowerSnapshot(cc, snapshotKey, &votingPowerSnapshot)
+	if err != nil {
+		return err
+	}
+
 	writeFunc()
 	return nil
 }
 
+func (k *Keeper) ClearVotingPowerSnapshot(ctx sdk.Context, avs string) error {
+	// calculate the time before which the snapshot should be cleared.
+	unbondingDuration, err := k.avsKeeper.GetAVSUnbondingDuration(ctx, avs)
+	if err != nil {
+		return operatortypes.ErrFailToClearVPSnapshot.Wrapf("ClearVotingPowerSnapshot: failed to get the avs unbonding duration, err:%s, avs:%s", err, avs)
+	}
+	epochInfo, err := k.avsKeeper.GetAVSEpochInfo(ctx, avs)
+	if err != nil {
+		return operatortypes.ErrFailToClearVPSnapshot.Wrapf("ClearVotingPowerSnapshot: failed to get the avs epoch information, err:%s, avs:%s", err, avs)
+	}
+
+	clearTime := ctx.BlockTime().Add(-epochInfo.Duration * time.Duration(unbondingDuration))
```

Comment:
I don't think that this is always correct because of the lag in sending / receiving IBC messages. Maybe a callback can be created (to be called by the AVS) to clear it when the AVS knows that a snapshot is not being used.

Reply:
It can be applied to AVSs that don’t need to address lag, such as Dogfood, and AVSs where all system logic is deployed on the Exocore chain. For appChain-type AVSs, we might consider using time or epoch numbers—obtainable from the callback—to replace `ctx.BlockTime()`. However, if we allow AVSs to change their epoch configuration, retrieving the block time may become challenging. 

We can adjust this part of the implementation after merging the app chain PR.